server/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java

```diff
@@ -109,19 +122,18 @@ default String describeTasks(List<T> tasks) {
         }
     }
 
-    record TaskResult(Exception failure) {
-        private static final TaskResult SUCCESS = new TaskResult(null);
-
-        public static TaskResult success() {
-            return SUCCESS;
+    record TaskResult(@Nullable ActionListener<ClusterState> taskListener, @Nullable Exception failure) {
+        public static TaskResult success(ActionListener<ClusterState> taskListener) {
+            return new TaskResult(Objects.requireNonNull(taskListener), null);
         }
 
         public static TaskResult failure(Exception failure) {
-            return new TaskResult(failure);
+            return new TaskResult(null, Objects.requireNonNull(failure));
         }
 
         public boolean isSuccess() {
-            return this == SUCCESS;
+            assert failure == null ^ taskListener == null;
```

Comment:
Should this assertion be moved to constructor?

```
public TaskResult {
  assert failure == null ^ taskListener == null;
}
```

Reply:
TIL records can have constructors :) 31331cbab6c